codex-rs/core/src/codex.rs

```diff
@@ -1474,6 +1478,101 @@ async fn submission_loop(
                 };
                 sess.send_event(event).await;
             }
+            Op::RunUserShellCommand { command } => {
+                // Spawn a cancellable one-off shell command task so we can process
+                // further Ops (e.g., Interrupt) while it runs.
+                let sess_clone = sess.clone();
+                let turn_context = Arc::clone(&turn_context);
+                let sub_id = sub.id.clone();
+                let handle = tokio::spawn(async move {
+                    // Announce a running task so the UI can show a spinner and block input.
+                    let event = Event {
+                        id: sub_id.clone(),
+                        msg: EventMsg::TaskStarted(TaskStartedEvent {
+                            model_context_window: turn_context.client.get_model_context_window(),
+                        }),
+                    };
+                    sess_clone.send_event(event).await;
+
+                    // Build a shell invocation in the user's default shell.
+                    let shell_invocation = sess_clone
+                        .user_shell
+                        // Why we pass a ["bash", "-lc", <script>] sentinel instead of the raw command:
+                        // - The shell adapter (core/src/shell.rs) first calls `strip_bash_lc`. When it sees this
+                        //   exact shape it extracts <script> and then builds the correct argv for the user shell
+                        //   (e.g., `/bin/zsh -lc "source ~/.zshrc && (<script>)"`).
+                        // - If we pass the whole command as a single string (e.g., ["cat Cargo.toml | wc -l"]) the
+                        //   adapter may quote it when joining/embedding, and shells can treat the entire value as a
+                        //   single program name or a single quoted token.
```

Comment:
Instead of relying on this quirk of `format_default_shell_invocation()`, we should probably just support this directly as a method of `Shell`?